### PR TITLE
Consensus should track core lifetime

### LINKED ISSUE #66: Consensus should track CoreLifetime of core
Consensus should track the lifetime of the Core which tells the high-level state of Core. 

The lifetime of a Core can be defined as: 
```
    CoreLifetime:undefined
    CoreLifetime:halted
    CoreLifetime:corrupted
    CoreLifetime:creation
    CoreLifetime:genesis
    CoreLifetime:active
```
Undefined: When core doesn't exist. 
Halted: Core stops functioning. 
Corrupted: Committee rejected the proposal/Pre-commit.
Genesis: When Core is `Opened` and no Pre-commit is done by Core. Once the first Pre-commit is received, the `CoreLifeTime` will change to Activated. 
Active: When Core is functioning correctly, the state can be `Opened` and `Precommitted`.
Creation: When Core is created.

### PR DESCRIPTION
**Pull request by 0xsarvesh**
fixes #66  
This PR adds functionality which enables consensus contracts to track CoreLifetime.

### CODE REVIEW

**Review by pgev — CHANGES_REQUESTED** [review]
@sarvesh-ost the proposal for CoreLifeTime has been slightly upgraded to include CoreLifeTime::Creation (https://hackmd.io/HPfjQEZKQh-lUCCP_NhYug?both#ConsensusCoreLifetime-and-CoreStatus). Please, update the corresponding ticket and implementation accordingly.

File: contracts/consensus/Consensus.sol [hunk 1]
```diff
@@ -25,8 +25,9 @@ import "../core/CoreI.sol";
 import "../core/CoreStatusEnum.sol";
 import "../reputation/ReputationI.sol";
 import "../proxies/MasterCopyNonUpgradable.sol";
+import "./CoreLifetime.sol";
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol]: please sort imports, move this to L20

1. external imports
space
2. local imports
3. other modules imports

sorted alphabethically

File: contracts/consensus/Consensus.sol [hunk 2]
```diff
@@ -301,6 +308,11 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
             "Committee formation blocksegment length must be in 256 most recent blocks."
         );
 
+        require(
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol]: perhaps move up to L293 as we first check active status of core; then check if precommit is present.

File: contracts/consensus/Consensus.sol [hunk 3]
```diff
@@ -635,8 +645,13 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
             _rootBlockHeight
         );
 
+        require(
+            coreLifetime[core] == CoreLifetime.undefined,
+            "Core lifetime status must be undefined"
+        );
         assignments[chainId] = core;
         anchors[chainId] = _anchor;
+        coreLifetime[core] = CoreLifetime.genesis;
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol]: CoreLifetime should be set to `creation` in `newCore`

File: contracts/consensus/Consensus.sol [hunk 4]
```diff
@@ -654,17 +669,18 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
     /* Internal functions */
 
     /**
-     * @notice Check if the core address is active.
+     * @notice Check if the core lifetime state is genesis or activated.
      * @param _core Core contract address.
      * Returns true if the specified address is a core.
      */
-    function isCore(address _core)
+    function isCoreInAction(address _core)
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol]: `isCoreRunning` might capture `genesis | active`

Comment (0xsarvesh) [reply contracts/consensus/Consensus.sol]: isCoreInAction renamed to isCoreActive which check `genesis` and `active`.

**Comment by 0xsarvesh** [comment]
This PR is available for review again.

File: contracts/consensus/Consensus.sol [hunk 5]
```diff
@@ -26,7 +27,7 @@ import "../core/CoreStatusEnum.sol";
 import "../reputation/ReputationI.sol";
 import "../proxies/MasterCopyNonUpgradable.sol";
 
-contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
+contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, ConsensusI {
```
Comment (pgev) [inline contracts/consensus/Consensus.sol]: Why do we inherit CoreLifetimeEnum and do not define it directly in contract?

File: contracts/consensus/Consensus.sol [hunk 6]
```diff
@@ -95,7 +96,7 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
     mapping(bytes20 /* chainId */ => bytes32 /* MetablockHash */) public metablockHeaderTips;
 
     /** Core statuses */
-    mapping(address /* core */ => CoreStatus /* coreStatus */) public coreStatuses;
+    mapping(address /* core */ => CoreLifetime /* coreLifetime */) public coreLifetime;
```
Comment (pgev) [inline contracts/consensus/Consensus.sol]: Suggestion:  coreLifetime -> coreLifetimes

File: contracts/consensus/Consensus.sol [hunk 7]
```diff
@@ -512,17 +522,21 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
...
+            coreLifetime[_core] == CoreLifetime.creation,
+            "Core lifetime status must be creation."
         );
 
         // Stake in reputation contract.
         reputation.stake(msg.sender, _withdrawalAddress);
 
         // Join in core contract.
-        CoreI(_core).joinDuringCreation(msg.sender);
+        uint256 coreValidatorCount = CoreI(_core).joinDuringCreation(msg.sender);
```
Comment (pgev) [inline contracts/consensus/Consensus.sol]: Proposal/Suggestion:

We can add `Core::minimumValidatorCountReached()` function.

File: contracts/core/Core.sol [hunk 8]
```diff
@@ -695,11 +695,14 @@ contract Core is MasterCopyNonUpgradable, ConsensusModule, MosaicVersion, CoreSt
...
      *
      * @param _validator A validator's address to join.
+     *
+     * @return The total count of joined validators.
      */
     function joinDuringCreation(address _validator)
         external
         onlyConsensus
         duringCreation
+        returns(uint256)
```
Comment (pgev) [inline contracts/core/Core.sol]: Here I am picky (it's still according to our CS :P), however, we can use named return variable to make interface easier.

File: contracts/consensus/Consensus.sol [hunk 9]
```diff
@@ -128,7 +129,7 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
     modifier onlyCore()
     {
         require(
-            isCore(msg.sender),
+            isCoreActive(msg.sender),
```
Comment (pgev) [inline contracts/consensus/Consensus.sol]: To me, isCoreActive() should return true only if core lifetime is set active (excluding here genesis).
I would also propose to change modifier name to `onlyActiveCore`.

**Review by pgev — COMMENTED** [review]
@sarvesh-ost thanks for changes. Some comments/suggestions.

File: contracts/consensus/Consensus.sol (hunk 4 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol]: I like @benjaminbollen 's proposal to have `isCoreRunning`. It's confusing to me to have `isCoreActive` function which checks that core is `active` or `genesis`. I would like to propose also to rename modifier to `onlyRunningCore`.

File: contracts/consensus/Consensus.sol (hunk 7 above)
Comment (benjaminbollen) [reply contracts/consensus/Consensus.sol]: we will update this in a later task

File: contracts/core/Core.sol (hunk 8 above)
Comment (benjaminbollen) [reply contracts/core/Core.sol]:
```
returns (uint256 validatorCount_, uint256 minValidatorCount_)
```

File: contracts/consensus/Consensus.sol (hunk 9 above)
Comment (benjaminbollen) [reply contracts/consensus/Consensus.sol]: `isCoreRunning` is explicitly proposed for this reason; it must be true for `genesis` and `active`

File: contracts/consensus/Consensus.sol (hunk 5 above)
Comment (benjaminbollen) [reply contracts/consensus/Consensus.sol]: It's unfortunately not possible in solidity I think; I am sad about that too

**Review by pgev — APPROVED** [review]

**Pull request merged**